tests/online/cert.c

```diff
@@ -0,0 +1,257 @@
+			cl_assert_equal_s(host, options->expect_host);
+		}
+	}
+
+	return (options ? options->return_code : GIT_ERROR);
+}
+
+void test_online_cert__certificate_invalid(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+	g_cert_options.return_code = GIT_ECERTIFICATE;
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+					 git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+void test_online_cert__certificate_invalid_ssh(void)
+{
+	if (!g_has_ssh)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.expect_host = "github.com";
+	g_cert_options.return_code = GIT_ECERTIFICATE;
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+					 git_clone(&g_repo, "ssh://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__certificate_valid(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+
+	cl_git_pass(git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__start_with_http(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+
+	cl_git_pass(git_clone(&g_repo, "http://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__passthrough_https(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_pass(git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_https_bad_certificate(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_fail_with(GIT_ERROR,
+					 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_https_other(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_EPEEL;
+
+	cl_git_fail_with(GIT_EPEEL,
+					 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &g_options));
```

Comment:
`GIT_EPEEL` is a very confusing error to use  here. `GIT_EUSER` would be much better as it's the one we have set up for people to return and have it passed all the way back.

Reply:
Ok, will fix. The intent was to make sure we preserved whatever the users' callback replied to us (hence the non-sensical `GIT_EPEEL`).